### ISSUE TITLE
[Feature Request]: Support for System JS as output format

### ISSUE THREAD

**Issue by jonbjk** [rolldown/rolldown#2387]
### What problem does this feature solve?

SystemJs is used in quite a few microfrontend architectures. Typically these use Rollup with output format "system", as build tool.
Are there any plans to support SystemJs output in Rolldown?

### What does the proposed API look like?

```
import { defineConfig } from 'rolldown';

export default defineConfig({
  input: 'src/dummy.js',
  output: {
    dir: 'dist',
    entryFileNames: `[hash].system.js`,
    format: 'system',
  },
});
```
Results in:
 ERROR  unimplemented: output.format: system

**Comment by hyf0** [comment]
Some notes:
- So far we don't have plan on supporting systemjs
- As a workaround, you could target to `esm` format and transpile the output to systemjs using other tools.
- We might provide a official/third-party plugin that does the above thing automatically

**Comment by izaakschroeder** [comment]
+1 from me. `systemjs` allows lower level access to module system plumbing than… say… import maps + ESM alone. As @hyf0 mentioned this is a pretty common thing for module federation which often needs to make low-level manipulation to the module resolution process. Also nice on the off chance you have to support ancient browsers that don't support ESM 😄

**Comment by 7086cmd** [comment]
My ideas:

> **TL;DR** SystemJS can be excluded from a separate `.rs` file, but the format requires specific actions and normalization. Using an internal plugin with the `VisitMut` trait can handle the format automatically.

1. SystemJS can be excluded from a separate `.rs` file. However, we should support the `system` format in the enum, as the `manifest` plugin requires specific actions to handle the SystemJS format. The format also introduces some keywords that require normalization, while normalizing the name in oxc via traverse is somewhat inconvenient at present.
2. We can utilize an internal plugin and implement the `VisitMut` trait to handle this. When the `format` is set to `system`, we can automatically enable the built-in plugin.
3. It appears that `@vitejs/plugin-legacy` requires the `system` format for legacy builds when polyfills are not included (related: https://github.com/vitejs/vite/blob/main/packages/plugin-legacy/src/index.ts#L437). Additionally, https://github.com/rolldown/vite/pull/43 mentions that the test for legacy requires the `system` format.
4. Rollup’s implementation of SystemJS is complex, which implies that implementing the `system` format internally would be somewhat burdensome. Using the oxc ast traverse is a more suitable approach, in my opinion.
5. I am uncertain whether `oxc-transformer` will support the `system` format in the future. However, the transformer may should pass all of Babel tests. Both sides can share the current status on this functionality.

**Comment by xavivars** [comment]
@hyf0 which type of feedback are you expecting to get?

**Comment by hyf0** [comment]
@xavivars Basically it's about why you need systemjs format. However, the current conclusion is rolldown itself won't support systemjs in a very long time.

So for anyone who wants systemjs output, I suggest exploring the way of 

> - As a workaround, you could target to esm format and transpile the output to systemjs using other tools.

I'll have this issue on our next team meeting and see if we had bandwidth to push it forward.

> - We might provide a official/third-party plugin that does the above thing automatically

**Comment by xavivars** [comment]
> Basically it's about why you need systemjs format

We need it (currently) because some browsers we need to support don't work with ESM import maps yet. And same for NodeJS. 

We're currently using rollup (which natively supports systemjs), and are hoping to migrate to rolldown to get the build-time benefits it brings

**Comment by hi-ogawa** [comment]
I just tested a simple idea of transforming each esm chunk into systemjs format via `renderChunk` hook. Here is my code 
https://github.com/hi-ogawa/reproductions/tree/main/rolldown-systemjs ([stackblitz](https://stackblitz.com/github/hi-ogawa/reproductions/tree/main/rolldown-systemjs?file=rolldown.config.system.js))

```js
export default defineConfig({
  input: ["./src/entry.js"],
  output: {
    dir: "dist",
    format: "esm",
    sourcemap: true,
  },
  plugins: [
    {
      name: "transform-chunk-systemjs",
      async renderChunk(code, _chunk) {
        const result = await babel.transformAsync(code, {
          babelrc: false,
          configFile: false,
          sourceMaps: true,
          plugins: [babelDynamicImport, babelSystemjs],
        });
        return { code: result.code, map: result.map };
      },
    },
  ],
});
```

It looks working for simple cases, but I'm not sure whether this approach works robustly on real projects. I'll test it later on rolldown vite legacy plugin use cases.

**Comment by tbrannam** [comment]
@hyf0 , hi-ogawa's work seems to technically work, but amplifies the performance difference between JS and Rust.    Transforming may be a reasonable approach, but doing it within the context of babel seems like it may discard many advantages of Rolldown.

> We might provide a official/third-party plugin that does the above thing automatically

What is required to have a 'native' transformation?

**Comment by tbrannam** [comment]
I could see a plugin using the plugin's parser (need to make sure that's actually present) - which I understand would give us the AST, but I think there wouldn't be a way to emit code from a transformed AST.    It would be I neat trick to be able to use the same mechanism as Rolldown to re-emit from a modified AST in a plugin.  Alternatively I think we'd need to consider using SWC as part of the plugin's implementation?

**Comment by IWANABETHATGUY** [comment]
> I could see a plugin using the plugin's parser (need to make sure that's actually present) - which I understand would give us the AST, but I think there wouldn't be a way to emit code from a transformed AST. It would be I neat trick to be able to use the same mechanism as Rolldown to re-emit from a modified AST in a plugin. Alternatively I think we'd need to consider using SWC as part of the plugin's implementation?我可以看到一个插件使用插件的解析器（需要确保它实际上存在）——我理解这将给我们提供 AST，但我认为没有办法从转换后的 AST 发出代码。如果能够使用与 Rolldown 相同的机制从修改后的 AST 在插件中重新发出代码，那将是一个很好的技巧。或者，我认为我们需要考虑将 SWC 作为插件实现的一部分？

Yeah, you could use swc to support systemjs format in plugin, it should be much faster than babel.

**Comment by ciddan** [comment]
> +1 from me. `systemjs` allows lower level access to module system plumbing than… say… import maps + ESM alone. As [@hyf0](https://github.com/hyf0) mentioned this is a pretty common thing for module federation which often needs to make low-level manipulation to the module resolution process. Also nice on the off chance you have to support ancient browsers that don't support ESM 😄

Pretty much the same feedback from me - in a large project I work on we have a home-rolled micro frontend framework that depends deeply on the SystemJS format, since it allows us to customize module loading beyond what's possible with vanilla esm imports + import maps. We also need to target certain environments that don't fully support native imports (pre-blink Edge).

**Comment by lisonge** [comment]
In my original project, several files are run through dynamic loading, but my final output can only consist of a single file. SystemJS allows multiple files to be merged into one without changing the runtime behavior.

**Comment by AjCu** [comment]
I've encountered the same issue with our microfrontends approach. We're currently testing migrating to Rolldown now that Vite 8 has adopted it as the default. Inspired by @hi-ogawa repro, I developed a [Vite plugin that transforms the output from ES to System](https://www.npmjs.com/package/vite-plugin-esm-to-systemjs), its working so far but probably needs more testing.

**Comment by esroyo** [comment]
@hyf0 I apologize for opening [PR #9612](https://github.com/rolldown/rolldown/pull/9612) without further discussion. I got the impression the decision was on hold (mostly) because of resource constraints.

The case for SystemJS as a format is well-established. There is nothing new to discuss in that area. It delivers ESM semantics via a userland implementation, which is both its main advantage and disadvantage: in its favor, being a library provides a level of flexibility that native ESM cannot (custom loading, import map declarations, etc.); against it, it becomes irrelevant if you can use ESM natively without further customizations.

Therefore, the discussion is more about the Rolldown team's willingness and capacity to support the format. That, as usual, is a trade-off. Is it worth paying the maintenance tax? I can't say. For me as a consumer, there are two selling points: keeping the promise of Rollup compatibility, and adding a competitive advantage over esbuild (which [won't support it](https://github.com/evanw/esbuild/issues/192)).

My team has been working with Rollup + SystemJS for over 8 years. We are a third-party SaaS and use it to deliver code by configuring custom import maps anytime and anywhere without polluting the host page &mdash;all while keeping the ESM mental model. We could find other ways to do the same, such as making custom builds and rewriting the module URLs, but none are as transparent, flexible, and nice as SystemJS.

To me, the advantages it brings outweigh the costs, because I don't feel it has a _huge_ cost &mdash;it is close enough to ESM semantics. However, I agree that the format is somewhat niche and won't necessarily be acknowledged as a treasured feature by the broader community.

I really hope the format can be supported in the Rolldown core. Thanks for your time, attention, and hard work.

---

P.S. Other advanced use cases we rely on include injecting multiple module bodies into a single response via named registers. This is useful for preloading or stubbing content while keeping the application agnostic to it (essentially bundling content at another layer in a federated module network). We also use it to protect ourselves from global scope pollution &mdash;for example, where an existing AMD `define` variable could alter a library's behavior (which we mitigate by injecting `var define;` into the intro). In the end, these are all highly specific use cases that are unlocked thanks to having the SystemJS format.

P.P.S. We have published a version that includes SystemJS support to the npm registry at `@thehotelsnetwork/rolldown@1.0.10`. Please note that it only ships bindings for `x86_64-unknown-linux-gnu`, `x86_64-unknown-linux-musl`, `aarch64-unknown-linux-gnu`, `aarch64-unknown-linux-musl`, and `wasm32-wasip1-threads`. We **do not** intend to maintain a long-term fork; this is simply a temporary workaround for our own use. _Use at your own risk_.

**Comment by tbrannam** [comment]
There seems to have been little support from upstream in addressing this gap.   Adding SystemJS opens up a path for migration for many project that have previously adopted SystemJS as a platform agnostic loader.   Support also gives a path forward for SingleSPA tool chain - where ESM isn't a viable solution due to browser support requirements by some consumers.

Esroyo's contribution seems to be an earnest attempt to address this gap.  Sad to see the effort rebuffed, while not providing guidance as to why it continues to be ignored here?

**Comment by izaakschroeder** [comment]
My original comment still stands as well, would love to have this feature for building more complex module flows that other systems cannot support.
